compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -1012,13 +1012,32 @@ where
                     // using more niches than just null (e.g., the first page of
                     // the address space, or unaligned pointers).
                     Variants::Multiple {
-                        tag_encoding: TagEncoding::Niche { untagged_variant, .. },
+                        tag_encoding:
+                            TagEncoding::Niche { untagged_variant, ref niche_variants, niche_start },
                         tag_field,
+                        ref variants,
                         ..
                     } if this.fields.offset(tag_field) == offset => {
-                        Some(this.for_variant(cx, untagged_variant))
+                        // When a non-null niche is passed in, we might pass an unaligned value,
+                        // so we need to adjust the alignment to an appropriate value.
+                        let align = variants
+                            .iter_enumerated()
+                            .filter_map(|(idx, variant)| {
+                                if untagged_variant == idx || variant.abi.is_uninhabited() {
+                                    None
+                                } else {
+                                    Some(idx)
+                                }
+                            })
+                            .fold(Align::MAX, |align, idx| {
+                                let niche = (idx.as_u32() as u128
```

Comment:
I think I know what this code is supposed to do because I asked for it above, but I can't figure it out by reading the source. What is `idx` here? I think it's a VariantIdx? I can't figure out why we would be doing arithmetic with one of those. Can you untangle this for me?

Reply:
> I think it's a VariantIdx?

Yes!

>  Can you untangle this for me?

I think the new changes should be much better now.